app/frontend/wallet/cardano-wallet.ts

```diff
@@ -29,73 +29,134 @@ function txFeeFunction(txSizeInBytes) {
   return Math.ceil(a + txSizeInBytes * b)
 }
 
-function estimateTxSize(txInputs, outAddress, hasChange, hasDonation) {
-  const txInputsSize = encode(new CborIndefiniteLengthArray(txInputs)).length
-  const outAddressSize = base58.decode(outAddress).length
+type UTxO = {
+  txHash: string
+  address: string
+  coins: number
+  outputIndex: number
+}
 
-  //size of addresses used by AdaLite
-  const ownAddressSize = 76
-  const donationAddressSize = base58.decode(ADA_DONATION_ADDRESS).length
+type Input = UTxO
+
+type Output = {
+  address: string
+  coins: number
+}
+
+// Estimates size of final transaction in bytes.
+// Note(ppershing): can overshoot a bit
+function estimateTxSize(inputs: Array<Input>, outputs: Array<Output>) {
+  // exact size for inputs
+  const preparedInputs = inputs.map(TxInputFromUtxo)
+  const txInputsSize = encode(new CborIndefiniteLengthArray(preparedInputs)).length
 
-  /*
-  * we assume that at most three outputs (destination, change and possibly donation address)
-  * will be present encoded in an indefinite length array
-  */
   const maxCborCoinsLen = 9 //length of CBOR encoded 64 bit integer, currently max supported
-  const txOutputsSize = hasChange
-    ? outAddressSize + ownAddressSize + maxCborCoinsLen * 2 + 2
-    : outAddressSize + maxCborCoinsLen + 2
 
-  const donationOutputSize = hasDonation ? donationAddressSize + maxCborCoinsLen : 0
+  const txOutputsSizes = outputs.map(
+    // Note(ppershing): we are conservative here
+    // FIXME(ppershing): shouldn't there be some +1 for the array encoding?
+    // Is it in maxCborCoinsLen?
```

Comment:
its an indefinite array, its cost is already included 2 rows below

Reply:
Note: this is about *inner* array, which is definite and has 2 elements [address, coins]